File: source/bxgenbb_help/src/decay0_fortran/funbeta2.f
c funbeta2.f 
c
c This file was extracted from the 'decay0' program by V.I. Tretyak
c
c Copyright 1995-2011 V.I. Tretyak
c
c This program is free software; you can redistribute it and/or modify
c it under the terms of the GNU General Public License as published by
c the Free Software Foundation; either version 3 of the License, or (at
c your option) any later version.
c 
c This program is distributed in the hope that it will be useful, but
c WITHOUT ANY WARRANTY; without even the implied warranty of
c MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
c General Public License for more details.
c 
c You should have received a copy of the GNU General Public License
c along with this program; if not, write to the Free Software
c Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
c 
c
	function funbeta2(E)
c Function to search the maximum of beta spectrum curve by tgold
c subroutine.
c VIT, 30.07.1992; 15.10.1995; 31.03.2006.
	common/const/pi,emass,datamass(50)
	common/parbeta/Zdtr,Qbeta
	common/parbeta2/kf,c1,c2,c3,c4
	funbeta2=0.
	if(E.gt.0.) then
c allowed spectrum
	   all=sqrt(E*(E+2.*emass))*(E+emass)*(Qbeta-E)**2*fermi(Zdtr,E)
c correction factor
	   w=E/emass+1.
	   pel=sqrt(w**2-1.)
	   pnu=(Qbeta-E)/emass
	   cf=1.
	   if(kf.eq.1) cf=pel**2+c1*pnu**2
	   if(kf.eq.2) cf=pel**4+c1*pel**2*pnu**2+c2*pnu**4
	   if(kf.eq.3) cf=pel**6+c1*pel**4*pnu**2+c2*pel**2*pnu**4+
     +                  c3*pnu**6
	   if(kf.eq.4) cf=pel**8+c1*pel**6*pnu**2+c2*pel**4*pnu**4+
     +                  c3*pel**2*pnu**6+c4*pnu**8
c spectrum with correction
	   funbeta2=all*cf
	endif
	return
	end
c
c end of funbeta2.f
c
c Local Variables: --
c mode: fortran --
c End: --

File: source/bxgenbb_help/src/decay0/funbeta2.cc
// -*- mode: c++; -*- 
// funbeta2.cc

#include <iostream>
#include <cmath>
#include <sstream>
#include <stdexcept>

#include <gsl/gsl_math.h>

#include <genbb_help/decay0/funbeta2.h>
#include <genbb_help/decay0/fermi.h>
#include <genbb_help/decay0/particle.h>

namespace genbb {
  namespace decay0 {
 
    /// \brief Function to search the maximum of beta spectrum curve by decay0_tgold subroutine
    double decay0_funbeta2(double E_, void * params_)
    {
      const parbeta2 * pars = static_cast<const parbeta2 *>(params_);
      const double & Zdtr  = pars->Zdtr;
      const double & Qbeta = pars->Qbeta;
      const double & c1    = pars->c1;
      const double & c2    = pars->c2;
      const double & c3    = pars->c3;
      const double & c4    = pars->c4;
      const int    & kf    = pars->kf;

      static const double emass = decay0_emass();
      double f = 0.0;
      if (E_ > 0.) {
        // allowed spectrum:
        double all = 
          std::sqrt(E_ * (E_ + 2. * emass)) 
          * (E_ + emass)
          * gsl_pow_2(Qbeta - E_) 
          * decay0_fermi(Zdtr, E_);
        // correction factor
        double w = E_ / emass + 1.;
        double pel = std::sqrt(gsl_pow_2(w) - 1.);
        double pnu = (Qbeta - E_) / emass;
        double cf = 1.;
        if (kf == 1) {
            cf =     gsl_pow_2(pel)
              + c1 * gsl_pow_2(pnu);
          }
        else if (kf == 2) {
            cf =     gsl_pow_4(pel)
              + c1 * gsl_pow_2(pel) * gsl_pow_2(pnu)
              + c2 *                  gsl_pow_4(pnu);
          }
        else if (kf == 3) {
            cf =     gsl_pow_6(pel)
              + c1 * gsl_pow_4(pel) * gsl_pow_2(pnu)
              + c2 * gsl_pow_2(pel) * gsl_pow_4(pnu)
              + c3 *                  gsl_pow_6(pnu);
          }
        else if (kf == 4) { 
            cf =     gsl_pow_8(pel)
              + c1 * gsl_pow_6(pel) * gsl_pow_2(pnu)
              + c2 * gsl_pow_4(pel) * gsl_pow_4(pnu)
              + c3 * gsl_pow_2(pel) * gsl_pow_6(pnu)
              + c4 *                  gsl_pow_8(pnu);
          }
        else if (kf > 4) {  
          std::ostringstream message;
          message << "genbb::decay0::decay0_funbeta2: " 
                  << "Invalid kf value ( "
                  << kf << ") !";
          throw std::logic_error (message.str());
        }
        // spectrum with correction
        f = all * cf;
      }
      return f;
    }

  } // end of namespace decay0 
} // end of namespace genbb 
